api/src/Validator/AllowTransition/AssertAllowTransitionsValidator.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+namespace App\Validator\AllowTransition;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\InvalidArgumentException;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+
+class AssertAllowTransitionsValidator extends ConstraintValidator {
+    public function __construct(public RequestStack $requestStack) {
+    }
+
+    public function validate($value, Constraint $constraint) {
+        if (!$constraint instanceof AssertAllowTransitions) {
+            throw new UnexpectedTypeException($constraint, AssertAllowTransitions::class);
+        }
+
+        $transitions = $constraint->getTransitions();
+        if (!$transitions->forAll(fn ($_, $elem) => self::isValidTransition($elem))) {
+            throw new InvalidArgumentException('All transitions must have a from and a to key');
+        }
+
+        $transitions = $transitions->map(fn ($elem) => new Transition($elem['from'], $elem['to']));
+
+        $allFrom = $transitions->map(fn (Transition $elem) => $elem->from)->toArray();
+        $allTo = $transitions->map(fn (Transition $elem) => $elem->to)->toArray();
+        $allTo = array_merge(...$allTo);
+        if (!(new ArrayCollection($allTo))->forAll(fn ($_, $to) => in_array($to, $allFrom, true))) {
+            throw new InvalidArgumentException('All to must appear in a from again');
+        }
+
+        $previousObject = $this->requestStack->getCurrentRequest()->attributes->get('previous_data');
+        $previousValue = $previousObject->{$this->context->getPropertyName()};
+
+        if ($value === $previousValue) {
+            return;
+        }
+
+        foreach ($transitions as $transition) {
+            if ($previousValue === $transition->from) {
+                $to = $transition->to;
+                if (!in_array($value, $to, true)) {
+                    $this->context->buildViolation('value must be one of {{ to }}, was {{ value }}')
+                        ->setParameter('{{ to }}', join(', ', $to))
+                        ->setParameter('{{ value }}', $value)
+                        ->addViolation()
+                    ;
+                }
+
+                return;
+            }
+        }
+
+        $this->context->buildViolation('previousValue must one of {{ from }}, was {{ previousValue }}')
```

Comment:
Ist diese Message hilfreich für Entwickler die die eCamp-API nutzen? Wie wäre es so:
```suggestion
        $this->context->buildViolation('This value was previously in an unexpected state, expected one of {{ from }}, but was {{ previousValue }}')
```
Und wenn wir `previousValue` wirklich rausgeben würde ich noch kurz überdenken ob das kein Informations-Leak ist (was passiert wenn wir irgendwann mal eine Entity haben die für den User nicht lesbar aber nur schreibbar ist?)

Reply:
Message habe ich angepasst.
Ich gebe den previous value noch heraus, macht das debuggen jetzt noch einfacher.
Können wir später wieder herausnehmen. Dies sollte so oder so nicht passieren, da wir zuerst die TO validieren.